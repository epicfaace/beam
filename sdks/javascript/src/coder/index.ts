/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Coder as Coder_, MergeStatus, AccumulationMode, OutputTime, ClosingBehavior, OnTimeBehavior, Trigger } from '../model/generated/beam_runner_api_pb';
import { FunctionSpec } from '../specs/function-spec';
import urns from '../model/urns';

class CoderSpec extends FunctionSpec {
  _urn() {
    return urns.StandardCoders.Enum.STRING_UTF8;
  }

  _payload() {
    return "";
  }
}

export default class Coder {
  spec: CoderSpec;

  constructor() {
    this.spec = new CoderSpec();
  }
  
  serialize() {
    const pb = new Coder_();
    pb.setSpec(this.spec.serialize());
    // pb.setComponentCoderIdsList([]); // TODO fix this
    return pb;
  }
}