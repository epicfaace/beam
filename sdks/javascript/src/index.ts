export { Pipeline } from './pipeline'
export { DoFn } from './specs/dofn'
export { Create, ParDo, PTransform } from './transforms'

// const x = 5 |> (x => x ** 2);
// console.error(x);

// export const Create = Create_
// export const ParDo = ParDo_
// export const PTransform = PTransform_