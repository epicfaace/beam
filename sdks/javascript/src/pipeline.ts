import {
  PTransform,
  Components,
  Pipeline as Pipeline_
} from '../model/generated/beam_runner_api_pb'

// TODO: support step names.
type Step = PTransform

export class Pipeline {
  /** Steps for this pipeline. */
  _steps: Step[] = []

  constructor(_props?: any) {
    // TODO: add pipelineoptions
    // TODO: CallableWrapperDOFn for map and flatmap
  }

  /**
   * Add a step to the pipeline.
   * @param {Step} Step.
   * @returns {Pipeline_} Serialized pipeline proto
   */
  pipe(step: Step) {
    this._steps.push(step)
    return this
  }

  /**
   * Serializes this pipeline to a runner api proto.
   * @returns {Pipeline_} Serialized pipeline proto
   */
  serialize() {
    const pipeline = new Pipeline_()
    const transform = new PTransform()
    transform.setUniqueName('unique name 1')

    const components = new Components()
    components.getTransformsMap().set('t', transform)
    pipeline.setComponents(components)
    const id = String(Math.random())
    pipeline.setRootTransformIdsList([id])

    // StandardPTransforms.Primitives.MAP_WINDOWS
    return pipeline
  }

  /**
   * Clones this pipeline.
   * @returns {Pipeline} Cloned pipeline
   */
  clone() {
    const p = new Pipeline()
    p._steps = [...this._steps]
    return p
  }

  /**
   * Builds this pipeline.
   * @returns {object} Object representing serialized pipeline
   */
  build() {
    return this.serialize().toObject()
  }
}
