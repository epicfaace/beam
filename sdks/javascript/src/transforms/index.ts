/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { Pipeline } from '../pipeline'
export { ParDo } from './pardo'
export { PTransform } from './ptransform'
export { Create } from './create'

// type PipelineArgs = any[]
// // type PipelineFn = (p: Pipeline) => (...args: PipelineArgs) => Pipeline;

// const pipelineify = (Cls: new (...args: PipelineArgs) => any) => (...args: PipelineArgs) => (
//   p: Pipeline
// ) => {
//   const cls = new Cls(...args)
//   return p.clone().pipe(cls)
// }

// export const PTransform_ = pipelineify(PTransform)
// export const ParDo_ = pipelineify(ParDo)
// export const Create_ = pipelineify(Create)
