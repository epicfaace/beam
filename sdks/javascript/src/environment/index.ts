/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import beam_runner_api_pb from '../model/generated/beam_runner_api_pb';
import { FunctionSpec } from '../specs/function-spec';
import urns from '../model/urns';
import { ApiServiceDescriptor } from '../model/generated/endpoints_pb';
import { PipelineContext } from '../pipeline/pipeline-context';

export class Environment extends FunctionSpec {

  /**
   * Serialize this object to a protobuf.
   * @return {FunctionSpec_} The generated protobuf.
   */
  serialize(_context: PipelineContext) {
    const spec = new beam_runner_api_pb.Environment();
    spec.setUrn(this._urn());
    spec.setPayload(this._payload() as Uint8Array);
    spec.addCapabilities("beam:coder:bytes:v1");
    spec.addCapabilities("beam:version:sdk_base:apache/beam_javascript_sdk:2.23.0.dev");
    // TODO: add proper capabilities
    return spec
  }
}

export class ExternalEnvironment extends Environment {
  url: string = "INVALID";

  _urn() {
    return urns.StandardEnvironments.EXTERNAL;
  }

  _payload() {
    const payload = new beam_runner_api_pb.ExternalPayload();
    const descriptor = new ApiServiceDescriptor();
    descriptor.setUrl(this.url);
    payload.setEndpoint(descriptor);
    return payload.serializeBinary();
  }
}