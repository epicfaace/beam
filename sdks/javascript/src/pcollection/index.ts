import { Pipeline } from 'core/pipeline';
import beam_runner_api_pb from "../model/generated/beam_runner_api_pb";

/**
 * Base class for PCollection.
 */
export class PValue {
  pipeline: Pipeline;
  name: string;
  constructor(pipeline: Pipeline) {
    this.pipeline = pipeline;
    this.name = pipeline.context.createUniqueRef(this);
  }

  // apply(...args) {
  //   args.splice(1, 0, this);
  //   return this.pipeline.apply(...args)
  // }
}

/**
 * A multiple values (potentially huge) container.
 */
export class PCollection extends PValue {
  serialize() {
    let pb = new beam_runner_api_pb.PCollection();
    pb.setUniqueName(this.name); // 14Create/Impulse.None
    pb.setCoderId("TODO"); // ref_Coder_BytesCoder_1
    pb.setIsBounded(beam_runner_api_pb.IsBounded.Enum.BOUNDED);
    pb.setWindowingStrategyId("TODO"); // ref_Windowing_Windowing_1
    return pb;
  }
}