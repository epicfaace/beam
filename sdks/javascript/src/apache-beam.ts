export { Pipeline } from './pipeline'
export { DoFn } from './specs/dofn'
import { Create_, ParDo_, PTransform_ } from './transforms'

export const Create = Create_
export const ParDo = ParDo_
export const PTransform = PTransform_
