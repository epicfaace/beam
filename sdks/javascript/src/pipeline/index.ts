import beam_runner_api_pb from '../model/generated/beam_runner_api_pb'
import { PTransform } from '../transforms/ptransform'
import { PValue, PBegin } from '../pcollection/pvalue'
import { PipelineContext } from './pipeline-context';
import { AppliedPTransform } from './applied-ptransform';
import { PipelineOptions } from './pipeline-options';
import { PipelineRunner } from '../runner/pipeline-runner';
import { DirectRunner } from '../runner/direct-runner';

export type PValueish = PValue | Pipeline

export class Pipeline {
  /** Steps for this pipeline. */
  transformsStack: AppliedPTransform[] = [];

  /** Set of transform labels applied to the pipeline. */
  appliedLabels: Set<string> = new Set();

  /** Pipeline context -- used for generating unique refs. */
  context: PipelineContext = new PipelineContext();

  /** Pipeline options */
  options: PipelineOptions;

  /** Pipeline runner */
  runner: PipelineRunner;

  constructor(runner?: PipelineRunner, options?: PipelineOptions) {
    this.runner = runner || new DirectRunner();;
    this.options = options || new PipelineOptions();

    const rootTransform = new AppliedPTransform(undefined, new PTransform(), "", []);
    rootTransform.fullLabel = this.context.createUniqueRef(rootTransform);
    this.transformsStack.push(rootTransform);
  }

  _currentTransform() {
    return this.transformsStack[this.transformsStack.length - 1];
  }

  _rootTransform() {
    return this.transformsStack[0];
  }

  begin() {
    return this.apply({ transform: new PTransform(), pvalueish: new PBegin(this)})
  }

  /**
   * Add a step to the pipeline.
   * @param {PTransform} PTransform.
   * @param {string} Label of the PTransform.
   * @param {PValue} Input for the PTransform, typically a PCollection.
   * @returns {Pipeline_} Serialized pipeline proto
   */
  apply({transform, label, pvalueish}: {transform: PTransform, label?: string, pvalueish?: PValueish}) {
    const fullLabel = this.context.createUniqueRef(this._currentTransform(), label);
    if (this.appliedLabels.has(fullLabel)) {
      throw new Error("label is already in use");
    }
    const inputs = transform.extractInputPValues(pvalueish);
    const appliedPTransform = new AppliedPTransform(this._currentTransform(), transform, fullLabel, inputs);

    this.appliedLabels.add(fullLabel);
    this._currentTransform().addPart(appliedPTransform);
    this.transformsStack.push(appliedPTransform);

    const pvalueResult = this.runner.apply({
      transform,
      pvalueish: pvalueish || this,
      options: this.options
    });

    appliedPTransform.addOutput(pvalueResult);

    return pvalueResult;
  }

  /**
   * Serializes this pipeline to a runner api proto.
   * @returns {Pipeline_} Serialized pipeline proto
   */
  serialize() {
    const pipeline = new beam_runner_api_pb.Pipeline();

    const components = new beam_runner_api_pb.Components();
    for (let transform of this.transformsStack) {
      components.getTransformsMap().set(transform.fullLabel, transform.serialize());
    }
    pipeline.setComponents(components);
    pipeline.setRootTransformIdsList([this._rootTransform().fullLabel]);

    // StandardPTransforms.Primitives.MAP_WINDOWS
    return pipeline
  }

  /**
   * Clones this pipeline.
   * @returns {Pipeline} Cloned pipeline
   */
  clone() {
    const p = new Pipeline()
    // p._steps = [...this._steps]
    return p
  }

  /**
   * Builds this pipeline.
   * @returns {object} Object representing serialized pipeline
   */
  build() {
    return this.serialize().toObject()
  }

  /**
   * Runs this pipeline.
   */
  async run() {
    return this.runner.runPipeline(this, this.options);
  }
}
