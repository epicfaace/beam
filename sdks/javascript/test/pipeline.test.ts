import { Pipeline, ParDo, DoFn } from '../src/index'

describe('Pipeline', () => {
  // it('no pipeline operator', () => {
  //   let p = new Pipeline();// |> Create(['a', 'b', 'c']);
  //   p = Create(['a', 'b', 'c'])(p)
  
  //   expect(p.serialize()).toMatchSnapshot();
  // });
  it('empty with single root transform', () => {
    let p = new Pipeline();
    expect(p.serialize().toObject()).toMatchSnapshot();
  });
  it('with one transform', () => {
    let p = new Pipeline();
    p.apply(new ParDo(new DoFn()), "custom label");
    expect(p.serialize().toObject()).toMatchSnapshot();
  });
  it('with multiple subtransforms', () => {
    let p = new Pipeline()
      .apply(new ParDo(new DoFn()), "A");
    
    p.apply(new ParDo(new DoFn()), "B-1")
      .apply(new ParDo(new DoFn()), "C-1");

    p.apply(new ParDo(new DoFn()), "B-2")
      .apply(new ParDo(new DoFn()), "C-2");
    expect(p.serialize().toObject()).toMatchSnapshot();
  });
  // it('with pipeline operator', () => {
  //   let p = new Pipeline() |> Create(['a', 'b', 'c']);
  //   console.log(p.build())
  
  //   expect(1 + 1).toEqual(2);
  // });
})
