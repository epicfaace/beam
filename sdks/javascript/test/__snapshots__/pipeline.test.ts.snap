// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pipeline empty with single root transform 1`] = `
Object {
  "components": Object {
    "codersMap": Array [],
    "environmentsMap": Array [],
    "pcollectionsMap": Array [],
    "transformsMap": Array [
      Array [
        "ref_AppliedPTransform_AppliedPTransform_1",
        Object {
          "displayDataList": Array [],
          "environmentId": "",
          "inputsMap": Array [],
          "outputsMap": Array [],
          "spec": Object {
            "payload": "[object Object]",
            "urn": "beam:transform:js:v1",
          },
          "subtransformsList": Array [],
          "uniqueName": "ref_AppliedPTransform_AppliedPTransform_1",
        },
      ],
    ],
    "windowingStrategiesMap": Array [],
  },
  "displayDataList": Array [],
  "requirementsList": Array [],
  "rootTransformIdsList": Array [
    "ref_AppliedPTransform_AppliedPTransform_1",
  ],
}
`;

exports[`Pipeline with multiple subtransforms 1`] = `
Object {
  "components": Object {
    "codersMap": Array [],
    "environmentsMap": Array [],
    "pcollectionsMap": Array [],
    "transformsMap": Array [
      Array [
        "ref_AppliedPTransform_AppliedPTransform_1",
        Object {
          "displayDataList": Array [],
          "environmentId": "",
          "inputsMap": Array [],
          "outputsMap": Array [],
          "spec": Object {
            "payload": "[object Object]",
            "urn": "beam:transform:js:v1",
          },
          "subtransformsList": Array [
            "ref_AppliedPTransform_A_2",
          ],
          "uniqueName": "ref_AppliedPTransform_AppliedPTransform_1",
        },
      ],
    ],
    "windowingStrategiesMap": Array [],
  },
  "displayDataList": Array [],
  "requirementsList": Array [],
  "rootTransformIdsList": Array [
    "ref_AppliedPTransform_AppliedPTransform_1",
  ],
}
`;

exports[`Pipeline with one transform 1`] = `
Object {
  "components": Object {
    "codersMap": Array [],
    "environmentsMap": Array [],
    "pcollectionsMap": Array [],
    "transformsMap": Array [
      Array [
        "ref_AppliedPTransform_AppliedPTransform_1",
        Object {
          "displayDataList": Array [],
          "environmentId": "",
          "inputsMap": Array [],
          "outputsMap": Array [],
          "spec": Object {
            "payload": "[object Object]",
            "urn": "beam:transform:js:v1",
          },
          "subtransformsList": Array [
            "ref_AppliedPTransform_custom label_2",
          ],
          "uniqueName": "ref_AppliedPTransform_AppliedPTransform_1",
        },
      ],
    ],
    "windowingStrategiesMap": Array [],
  },
  "displayDataList": Array [],
  "requirementsList": Array [],
  "rootTransformIdsList": Array [
    "ref_AppliedPTransform_AppliedPTransform_1",
  ],
}
`;
